### PR TITLE
[Replaced with #125014] Check type of device_type for #124738

### LINKED ISSUE #124738: `torch.autocast` produces confusing error message when passing `torch.device`
### 🐛 Describe the bug

When using `torch.autocast`, the device argument does not accept instances of `torch.device` (as expected from the documentation). The fact that it errors is expected, as the documentation states that this is unsupported, but the error message produced is not informative/confusing.

```python
import torch

device = torch.device('cuda') # or 'cpu'

# Note that we should have passed device.type here, so the fact that this errors isn't itself a bug.
with torch.autocast(device_type=device):
    pass
```

This code block will return:

```RuntimeError: User specified an unsupported autocast device_type 'cuda'``` (or `'cpu'` if used above)

which is fairly uninformative/confusing, given that 'cuda' is actually an allowed `device_type`. 

Suggestion: perhaps print the type of the `device_type` if it isn't already a string (or complain about the type instead of the `str` representation)? If needed, I can submit a simple PR to do this.

The error arises from this line which just prints the `str` representation of the passed argument: https://github.com/pytorch/pytorch/blob/47330ca13321a42d4f1e75f091e17183227ae073/torch/amp/autocast_mode.py#L241

### Versions

```
Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.2
[pip3] torch_geometric==2.5.2
[pip3] torchaudio==2.2.2
[pip3] torchvision==0.17.2
[pip3] triton==2.2.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.2.2                    pypi_0    pypi
[conda] torch-geometric           2.5.2                    pypi_0    pypi
[conda] torchaudio                2.2.2                    pypi_0    pypi
[conda] torchvision               0.17.2                   pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi
```

(Note that I didn't include my system details since this is system-invariant, but let me know if you do want those).

cc @svekars @brycebortree @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by alat-rights**
Fixes #124738

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

**Comment by alat-rights** [comment]
Interesting, I've signed the CLA already.

File: torch/amp/autocast_mode.py [hunk 1]
```diff
@@ -198,8 +198,12 @@ def __init__(
             # TODO: support get_autocast_gpu/cpu_dtype
             assert dtype is not None
             return
-        self.device = device_type
         self.custom_backend_name = torch._C._get_privateuse1_backend_name()
+        self.device = device_type
+        if type(self.device) != str:
```
Comment (soulitzer) [inline torch/amp/autocast_mode.py:203]: nit:
```suggestion
        if not isinstance(self.device, str):
```

File: torch/amp/autocast_mode.py [hunk 2]
```diff
@@ -198,8 +198,12 @@ def __init__(
...
+            raise RuntimeError(
```
Comment (soulitzer) [inline torch/amp/autocast_mode.py:204]:
```suggestion
            raise ValueError(
```

File: torch/amp/autocast_mode.py [hunk 3]
```diff
@@ -198,8 +198,12 @@ def __init__(
...
+                f"Expected `device_type` of type `str`, NOT `{type(self.device)}`"
```
Comment (soulitzer) [inline torch/amp/autocast_mode.py:205]:
```suggestion
                f"Expected `device_type` of type `str`, got: `{type(self.device)}`"
```

**Review by soulitzer — COMMENTED** [review]
Thanks! Added a few small nits. Also a test would be nice.

**Comment by soulitzer** [comment]
> Interesting that EasyCLA is failing, I've signed the CLA already (just now).

this should be a known issue with the infra right now

**Comment by alat-rights** [comment]
I'm having some trouble with the `autocast` unit tests locally.
image

The specific failures also do not make much sense to me.

I'm working on fixing.

**Comment by alat-rights** [comment]
Unit tests are still failing for me locally.

I think I might retry the unit tests off `main` (without my changes) to try and isolate

image

**Comment by alat-rights** [comment]
Fixed! All tests pass now.

image

**Review by soulitzer — APPROVED** [review]
LGTM, thanks for the quick update

**Comment by soulitzer** [comment]
Hmm maybe the CLA failures are legit, the commits appear to be submitted by "=', e.g. I see the message: "= committed 11 hours ago", etc. Any idea why that might be the case?

(If it makes things easier, feel free to file a new PR and tag me there.)

**Comment by alat-rights** [comment]
Hi!

Yeah, my git config was messed up.

Sure. If it's blocking, I can open a new PR.

Do you know what exactly the CLA bot requires in the commit metadata? Any chance I could save this with a force push or similar?

**Comment by alat-rights** [comment]
Oh no... Sorry my git-fu needs some work.

I'll try to save it! >.<

**Pull request closed without merging**